proposals/windows-mdm-commands-tables.md

```diff
@@ -0,0 +1,118 @@
+> These commands contain operations to be performed on a given device. 
+> They don't yet have the MessageID, SessionID, and CmdId information as this data gets populated during the response 
+> to a given MDM management request.
+
+### `windows_mdm_commands`
+
+> The windows_mdm_commands table represents the stateful commands that were actually sent over the wire. 
+> These commands include the MessageID, SessionID, and CmdId as they were sent to the device. 
+> They are stored in the commands table for error code tracking purposes. 
+> Additionally, this table can contain commands that were not part of the windows_mdm_pending_commands table and 
+> were generated by the Fleet server in response to an incoming MDM command.
+
+## Proposal
+
+I think it's fundamental we have:
+
+- A clear queue model
+- A single source of truth for commands
+- The status of a command at any point in time easily available
+- The raw XML we sent to the device
+- The raw XML we received as a response
+
+A rough version of the schema (many missing fields from the original) could be:
+
+```sql
+CREATE TABLE windows_mdm_commands (
+    -- managed and generated by Fleet
+    command_uuid    VARCHAR(127) NOT NULL,
+
+    -- raw XML
+    raw_command      MEDIUMTEXT   NOT NULL,
+
+    -- command_id specifies a unique command identifier for the SyncML message (Section 2.2.3.2 on MS-MDM spec)
+    command_id      VARCHAR(255) NOT NULL,
+
+    -- cmd_verb is e.g. Add, Exec, Get, etc.    
+    cmd_verb        VARCHAR(20)  NOT NULL,
+
+    created_at      TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    updated_at      TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+
+    PRIMARY KEY (command_uuid, device_id)
+)
+
+CREATE TABLE windows_mdm_command_queue (
+    device_id    VARCHAR(255) NOT NULL,
+    command_uuid VARCHAR(127) NOT NULL,
+
+    active   BOOLEAN NOT NULL DEFAULT 1,
+    priority TINYINT NOT NULL DEFAULT 0,
+
+    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+
+    PRIMARY KEY (device_id, command_uuid),
+
+    INDEX (priority DESC, created_at),
+
+    FOREIGN KEY (device_id)
+        REFERENCES mdm_windows_enrollments (id)
+        ON DELETE CASCADE ON UPDATE CASCADE,
+
+    FOREIGN KEY (command_uuid)
+        REFERENCES commands (command_uuid)
+        ON DELETE CASCADE ON UPDATE CASCADE
+)
+
+CREATE TABLE windows_command_results (
+    device_id    VARCHAR(255) NOT NULL,
+    command_uuid VARCHAR(127) NOT NULL,
+    status       VARCHAR(31)  NOT NULL,
+
+    -- raw response, as received from the client
+    result       MEDIUMTEXT   NOT NULL,
+
+    -- this is the error code returned from the MDM Client Stack through Status/Results cmds
+    -- empty field means that no response has been received for the command
+    rx_error_code   VARCHAR(20) NOT NULL,
```

Comment:
I think the comment about "empty meaning no response received" would not apply anymore in this schema, right? Since this is the results table, no response received means no corresponding row in that table.

Reply:
There are two type of protocol command responses: `Status` and `Results`. 

`Status` contains the error code for a given protocol command. This is mandatory and the receiving party should send an error code for each protocol command received.

On the other hand, `Results` only applies to command that actually requires returning data (Get for example)

It could be that `Status` arrives first and then it arrives the `Results` command (see example above). So it should be fine to keep this field empty. It might help to pinpoint issues